tools/clang/unittests/HLSLExec/LongVectors.h

```diff
@@ -199,62 +206,189 @@ void FillLongVectorDataFromShaderBuffer(MappedData &ShaderBuffer,
+  for (size_t i = 0; i < Length; i++) {
+    if (Values[i].OpTypeString == OpTypeString) {
+      return static_cast<T>(Values[i].OpTypeValue);
+    }
+  }
+
+  LogErrorFmtThrow(L"Invalid LongVectorOpType string: %s",
+                   OpTypeString.c_str());
+
+  if (std::is_same_v<T, LongVectorBinaryOpType>)
+    return static_cast<T>(LongVectorBinaryOpType_EnumValueCount);
+  else if (std::is_same_v<T, LongVectorUnaryOpType>)
+    return static_cast<T>(LongVectorUnaryOpType_EnumValueCount);
+}
+
+static const LongVectorOpTypeStringToEnumValue
+    LongVectorBinaryOpTypeStringToEnumMap[] = {
+        {L"LongVectorBinaryOpType_ScalarAdd", LongVectorBinaryOpType_ScalarAdd},
+        {L"LongVectorBinaryOpType_ScalarMultiply",
+         LongVectorBinaryOpType_ScalarMultiply},
+        {L"LongVectorBinaryOpType_Multiply", LongVectorBinaryOpType_Multiply},
+        {L"LongVectorBinaryOpType_Add", LongVectorBinaryOpType_Add},
+        {L"LongVectorBinaryOpType_Min", LongVectorBinaryOpType_Min},
+        {L"LongVectorBinaryOpType_Max", LongVectorBinaryOpType_Max},
+        {L"LongVectorBinaryOpType_ScalarMin", LongVectorBinaryOpType_ScalarMin},
+        {L"LongVectorBinaryOpType_ScalarMax", LongVectorBinaryOpType_ScalarMax},
+};
+
+static_assert(_countof(LongVectorBinaryOpTypeStringToEnumMap) ==
+                  LongVectorBinaryOpType_EnumValueCount,
+              "LongVectorBinaryOpTypeStringToEnumMap size mismatch. Did you "
+              "add a new enum value?");
+
+LongVectorBinaryOpType
+GetLongVectorBinaryOpType(const std::wstring &OpTypeString) {
+  return GetLongVectorOpType<LongVectorBinaryOpType>(
+      LongVectorBinaryOpTypeStringToEnumMap, OpTypeString,
+      std::size(LongVectorBinaryOpTypeStringToEnumMap));
+}
+
+enum LongVectorUnaryOpType {
+  LongVectorUnaryOpType_Clamp,
+  LongVectorUnaryOpType_Initialize,
+  LongVectorUnaryOpType_EnumValueCount
 };
 
+static const LongVectorOpTypeStringToEnumValue
+    LongVectorUnaryOpTypeStringToEnumMap[] = {
+        {L"LongVectorUnaryOpType_Clamp", LongVectorUnaryOpType_Clamp},
+        {L"LongVectorUnaryOpType_Initialize", LongVectorUnaryOpType_Initialize},
+};
+
+static_assert(_countof(LongVectorUnaryOpTypeStringToEnumMap) ==
+                  LongVectorUnaryOpType_EnumValueCount,
+              "LongVectorUnaryOpTypeStringToEnumMap size mismatch. Did you add "
+              "a new enum value?");
+
+LongVectorUnaryOpType
+GetLongVectorUnaryOpType(const std::wstring &OpTypeString) {
+  return GetLongVectorOpType<LongVectorUnaryOpType>(
+      LongVectorUnaryOpTypeStringToEnumMap, OpTypeString,
+      std::size(LongVectorUnaryOpTypeStringToEnumMap));
+}
+
+template <typename T> std::vector<T> GetInputValueSetByKey(std::wstring Key) {
+
+  if constexpr (std::is_same_v<T, HLSLBool_t>)
+    return std::vector<T>(DefaultInputValueSet_bool.at(Key).begin(),
+                          DefaultInputValueSet_bool.at(Key).end());
```

Comment:
This seems convoluted to me.  If there's a good reason for it to be this way then a comment would be good.

The value type of these sets is already a vector of the right type?

```suggestion
  if constexpr (std::is_same_v<T, HLSLBool_t>)
    return DefaultInputValueSet_bool.at(Key);
```

I think you'd write it this way if the element types didn't match and you wanted to reconstruct each element...but why not just have them be the right type in the map to start with so they don't need to be converted again here?

Reply:
You're right. I was hitting an issue with the implicit copy. But that's only an issue for the HLSL* types. I'll clean that up by making the map the right type to begin with.